design/volume-group-snapshot.md

```diff
@@ -0,0 +1,428 @@
+
+## Goals
+- Ensure that multiple related volumes are snapshotted simultaneously, preserving consistency for stateful applications via VolumeGroupSnapshots(VGS) API.
+- Integrate VolumeGroupSnapshot functionality into Velero’s existing backup and restore workflows.
+- Allow users to opt in to volume group snapshots via specifying the group label.
+
+## Non-Goals
+- The proposal does not require a complete overhaul of Velero’s CSI integration, it will extend the current mechanism to support group snapshots.
+- No any changes pertaining to execution of Restore Hooks
+
+## High-Level Design
+
+Backup workflow:
+- Accept the label to be used for VGS from the user:
+    - Accept the label from the user, we will do this in 3 ways:
+        - Firstly, we will have a hard-coded default label key like `velero.io/volume-group-snapshot` that the users can directly use on their PVCs.
+        - Secondly, we will let the users override this default VGS label via a velero server arg, `--volume-group-nsaphot-label-key`, if needed.
+        - And Finally we will have the option to override the default label via Backup API spec, `backup.spec.volumeGroupSnapshotLabelKey`
+        - In all the instances, the VGS label key will be present on the backup spec, this makes the label key accessible to plugins during the execution of backup operation.
+    - This label will enable velero to filter the PVC to be included in the VGS spec.
+    - Users will have to label the PVCs before invoking the backup operation.
+    - This label would act as a group identifier for the PVCs to be grouped under a specific VGS.
+    - It will be used to collect the PVCs to be used for a particular instance of VGS object.
+
+- Changes to the Existing PVC ItemBlockAction plugin:
+    - Currently the PVC IBA plugin is applied to PVCs and adds the RelatedItems for the particular PVC into the ItemBlock.
+    - At first it checks whether the PVC is bound and VolumeName is non-empty.
+    - Then it adds the related PV under the list of relatedItems.
+    - Following on, the plugin adds the pods mounting the PVC as relatedItems.
+    - Now we need to extend this PVC IBA plugin to add the PVCs to be grouped for a particular VGS object, so that they are processed together under an ItemBlock by Velero.
+        - First we will check if the PVC that is being processed by the plugin has the user specified VGS label.
+        - If it is present then we will execute a List call in the namespace with the label as a matching criteria and see if this results in any PVCs (other than the current one).
+        - If there are PVCs matching the criteria then we add the PVCs to the relatedItems list.
+        - This helps in building the ItemBlock we need for VGS processing, i.e. we have the relevant pods and PVCs in the ItemBlock.
+
+- Updates to CSI PVC plugin:
```

Comment:
For CSI snapshot backup (without data mover), let's also clarify how the unused VGS objects are deleted from the cluster after backup. cc @blackpiglet

Reply:
I agree. The VolumeGroupSnapshot-related resources generated during the backup should be deleted in the backup process.

For the VolumeGroupSnapshot, it's very similar to the VolumeSnapshot, it's a namespace-scoped resource. We should try to avoid the namespace deletion causing the snapshot lose.